src/Random/InverseTransformSampling.h

```diff
@@ -0,0 +1,214 @@
+            for (unsigned d = 0; d < Dim; ++d) {
+                nr_m[d] = dist.cdf(regions(rank)[d].max(), d) - dist.cdf(regions(rank)[d].min(),d);
+                dr_m[d]   = dist.cdf(rmax[d],d) - dist.cdf(rmin[d],d);
+                umin[d] = dist.cdf(regions(rank)[d].min(),d);
+                umax[d] = dist.cdf(regions(rank)[d].max(),d);
+            }
+            T pnr = std::accumulate(nr_m.begin(), nr_m.end(), 1.0, std::multiplies<T>());
+            T pdr = std::accumulate(dr_m.begin(), dr_m.end(), 1.0, std::multiplies<T>());
+            
+            T factor = pnr / pdr;
+            nlocal_m      = factor * ntotal;
+            
+            unsigned int ngobal = 0;
+            MPI_Allreduce(&nlocal_m, &ngobal, 1, MPI_UNSIGNED_LONG, MPI_SUM,
+                          ippl::Comm->getCommunicator());
+            
+            int rest = (int)(ntotal - ngobal);
+            
+            if (rank < rest) {
+                ++nlocal_m;
+            }
+        }
+
+        template <class GeneratorPool>
+        struct fill_random {
+            //using view_type  = typename ippl::detail::ViewType<Tt, 1>::view_type;
+            using value_type = T; //typename Tt::value_type;
+            Distribution dist;
+            view_type x;
+            GeneratorPool rand_pool;
+            Vector<T, Dim> umin_m;
+            Vector<T, Dim> umax_m;
+            KOKKOS_FUNCTION
+            fill_random(Distribution dist_, view_type x_, GeneratorPool rand_pool_, Vector<T, Dim> umin_, Vector<T, Dim> umax_)
+            : dist(dist_)
+            , x(x_)
+            , rand_pool(rand_pool_)
+            , umin_m(umin_)
+            , umax_m(umax_){}
+            KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const {
+                typename GeneratorPool::generator_type rand_gen = rand_pool.get_state();
+                
+                value_type u = 0.0;
+                for (unsigned d = 0; d < Dim; ++d) {
+                    u       = rand_gen.drand(umin_m[d], umax_m[d]);
+                    
+                    // first guess for Newton-Raphson
+                    x(i)[d] = dist.estimate(u, d);
+                    
+                    // solve
+                    ippl::random::detail::NewtonRaphson<value_type> solver;
+
+                    solver.solve(dist, d, x(i)[d], u);
+
+                    rand_pool.free_state(rand_gen);
+                }
+            }
+        };
+        KOKKOS_INLINE_FUNCTION unsigned int getLocalNum() const { return nlocal_m; }
+        void generate(view_type view, Kokkos::Random_XorShift64_Pool<> rand_pool64) {
+            Kokkos::parallel_for(nlocal_m, fill_random<Kokkos::Random_XorShift64_Pool<>>(dist, view, rand_pool64, umin, umax));
+            Kokkos::fence();
+        }
+    private:
+        unsigned int nlocal_m;
+    };
+
+    template <typename T, unsigned Dim, unsigned DimP, typename PDF, typename CDF, typename ESTIMATE>
+    class Distribution {
```

Comment:
The namespace is called `random`, the headerfile Random.h and inside we have `Distribution`. Does this all add up?
This needs to be clarified

Reply:
The name space is random, which I think is chosen to be similar to other libraries I guess.
Distribution class is now defined in Distribution.h.
I renamed random.h to Utility.h to include useful functions and functors, like NewtonRaphson.
Everything related to Normal distribution is now in NormalDistribution.h.